### ISSUE TITLE
Find MyClass.class reference.

### ISSUE THREAD

**Issue by mika-koivusaari** [TNG/ArchUnit#131]
I'm trying to find unused classes in a project and ArchUnit doesn't seem to find MyClass.class references, would it be possible to add this feature? or is it just a matter of me not knowing how to do it properly?

Example:
```
public class MyClass {
    ...
}

public class MySecondClass{
    public void methodA(){
        methodThatTakesClassAsParameter(MyClass.class);
    }
    ...
}
```
MySecondClass references MyClass.

**Comment by codecholeric** [comment]
I think if the only way `MyClass` is referenced is indeed as an argument to a generic `foo(Class<?> type)` method, ArchUnit will not detect this at the moment.
The problem is, that `MyClass` then only appears in the constant pool of `MySecondClass`. `foo(..)` has a generic signature, depending on `Class.class` and not on `MyClass.class`, and the access to `MyClass.class` is in fact only an access to the constant pool

```
void foo();
    descriptor: ()V
    flags:
    Code:
      stack=2, locals=1, args_size=1
         0: aload_0
         1: ldc           #2                  // class MyClass
         3: invokespecial #3                  // Method bar:(Ljava/lang/Class;)V
         6: return
      LineNumberTable:
        line 5: 0
        line 6: 6
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0       7     0  this   LFoo;
```

ArchUnit uses the ASM library under the hood and I'm not exactly sure, if it's possible to access this detail of information. What I can say for sure is that it is missing from ArchUnit at the moment.
I can keep this issue open for further research, to analyze if it's possible to retrieve this info using ASM. If not we're probably out of luck (unless I would start to write my own low level byte code parser which I'm not willing to do just yet :wink:). But in the end ArchUnit can only analyse static properties. If your only reference to `MyClass` is as an argument to a method taking a `Class<?>`, this is close to the limits. Similar to `Class.forName("MyClass")`, which ArchUnit can't detect either. It pretty much means that the only way you interact with `MyClass` is via reflection / dynamically.

Side note: IMHO unused code is more of a style thing than a matter of architecture. At least if we consider architecture as "things that are hard to change" :wink: Unused code I can delete at any time in the blink of an eye without further consequences. So maybe it might be good enough to run your IDE inspection once in a while, locate those unused classes and remove them. In the end one can never be completely sure by just looking at static code, if a class is really unnecessary. If it's loaded via reflection / some classpath magic, it might still be relevant even if unreferenced. So in the end there will always be some manual process involved to look at classes and decide if they can really go...

**Comment by mika-koivusaari** [comment]
Thanks.

Is there some tutorials or tips for reading if one would like to contribute something like this for ArchUnit? I guess ASM library would be one.

Side note, true, not really an architecture think, but it's nice if tools are flexible and suit tasks not originally intended.  And I like the fact I can code check for example configuration files that reference classes. I think this would be more of a head up to the committer that this might be unused after your change, check it out. As "run manually every now and then" usually means never after the first time:)

**Comment by codecholeric** [comment]
Yeah, in this case it would be the ASM library itself and the `com.tngtech.archunit.core.importer.JavaClassProcessor` within the ArchUnit code as the interface to ASM. The most important integrative test of the import is the `com.tngtech.archunit.core.importer.ClassFileImporterTest`.
We're happy about any contribution :smiley: (and I'm happy to help in any way I can of course)
You can compare https://github.com/TNG/ArchUnit/blob/master/CONTRIBUTING.md for further info.

And yes, I'm a big fan of automation, don't get me wrong :wink: I just think that luckily, on the contrary to managing dependencies, unused code is fairly easy to remove at any point. (nevertheless one has to really do this, and yes, if it's a manual, optional task, often enough nobody does it :wink:)

**Comment by codecholeric** [comment]
@mika-koivusaari are you still looking into this? Otherwise I would close the issue for now since I don't see any way to implement this within the current infrastructure.

**Comment by mika-koivusaari** [comment]
@codecholeric No, you where right about it requiring changes in asm library.

**Comment by codecholeric** [comment]
Okay, I'll close this issue then for now.

**Comment by timtebeek** [comment]
Came across this thread while searching for a way to detect unused classes, as discussed here as well.
Found a way to detect them, even with the limitations discussed here and in other issues:
https://github.com/timtebeek/archunit-unreferenced

Hope it's helpful for others that come across this thread, as it is somewhat possible still. At least I was able to delete a good few thousand lines of code with it. :)
